### DISCUSSION TITLE
Finding out what resource triggered a job

### DISCUSSION THREAD

**Discussion by MuaawiaArshad-okta** [concourse/concourse#6663] [Help & Support]
Is there a way to tell what resource triggered a job. 
In particular I want to be able to notify if the job was triggered by a commit, or if it was triggered by something like docker image changing.

**Comment by chenbh** [comment]
There's no way to figure it out from within a task/other steps. The only place this info is revealed is through the web UI where the step name is in a different colour.
![Screen Shot 2021-03-15 at 10 24 28 AM](https://user-images.githubusercontent.com/17577073/111169152-17194080-8579-11eb-848c-8ab03dc3d691.png)

**Reply by MuaawiaArshad-okta** [reply]
Got it. is there some api that I can query to figure this out. Like use whatever the UI uses to color code this info?

**Reply by chenbh** [reply]
It hits the `/api/v1/builds/<build id>/resources` which returns the inputs and outputs for a build. The `first_occurance` field is set to true if it's a different version from the last build. 
https://github.com/concourse/concourse/blob/28153d0cb8dd6f670d09a2001bd7b27d74ddf217/atc/build_inputs_outputs.go#L3-L18

I just realized this doesn't fully answer which resource triggered a build, since non-triggering resources also shows up as new versions. You could theoretically do something super fancy by grabbing the build plan from `/api/v1/builds/<build id>/plan`, parsing it and filter on `get` steps with `trigger: true`, and then mapping that to the build input to figure out which were the actual triggers.

**Reply by MuaawiaArshad-okta** [reply]
Thanks for answering. :) 
Btw is there a reason something like this isn't natively available inside tasks? Or are there plans to get that information in at some point?

**Reply by chenbh** [reply]
I'll pick apart your question a little bit because it's actually kinda loaded.

> Why is something like this not natively available

We simply never had a need or feature request for it. I would love to learn more about why you care about why that information is important to you. Is it simply nice to get a slack notification that a code change upstream is triggering a new build (if you are using Concourse for CI)? Is it for audit reasons like making sure a newly published docker image has been blessed by security / compliance? Is it to output to some external system to keep track of events?

> Why is something like this not available inside tasks

As one of the design principles, Concourse intentionally tries to decouple task scripts from Concourse principles (like builds and resources), this is the reason why a lot of [build metadata are only available to resources](https://concourse-ci.org/implementing-resource-types.html#resource-metadata). The recommended way of doing notifications would be to model it as a resource and `put`-ing to it.

**Reply by j-petit** [reply]
Hi just stumbled upon this thread because we have the exact same need: getting somehow the information which resource triggered our pipeline.

Let me quickly explain our use case:
We have a microservice based application where we run integration tests in a pipeline triggered by the registry-image resource. Whenever a new image is pushed, we run the integration tests. Now, I would like to distribute back the test information to the actual code changes: as we are having one repository for each component and image, I'd like to update the build status in Bitbucket of this specific commit in the right repository.

I hope this clearly explains why this is useful. Do you see any way to achieve this? I understand that build metadata is only available to resources but even if I create a custom resource it's not clear to me how to do what we want. 

Thanks!
